### ISSUE TITLE
Read/Write methods not working as expected

### ISSUE THREAD

**Issue by EduardoNigro** [trezor/cython-hidapi#124]
Hello,

I have been working in porting some code that I wrote in **MATLAB** (to communicate with the LEGO EV3 device via USB) to Python, with the intention to make it more accessible to others.

However, my Python implementation using **_cython-hidapi_** seems to be stuck with the last message that was sent when using the **MATLAB** implementation of the same code. In other words, it seems that although messages appear to be exchanged between the PC and the device, something isn't quite right.

The idea is that a message (direct command) is sent to the EV3 device and a reply with data or an acknowledgement is sent back to the PC. This way, one can assure that no packets were lost during communication. A simple example code that: opens the device, sends two distinct messages (and prints the corresponding replies), and closes the device is shown below.

> import hid
> 
> #### # Opening EV3 device
> ID_VENDOR_LEGO = 0x0694
> ID_PRODUCT_EV3 = 0x0005
> device = hid.device()
> device.open(ID_VENDOR_LEGO, ID_PRODUCT_EV3)
> 
> #### # Getting EV3 brick serial number
> device.write(b'\n\x00\x00\x00\x00\x10\x00\xd3\x0c\x02\x10`')
> reply = device.read(1024)[0:24]
> print(reply)
> 
> #### # Getting EV3 brick name
> device.write(b'\t\x00\x00\x00\x00\x08\x00\xd3\r\x08`')
> reply = device.read(1024)[0:24]
> print(reply)
> 
> #### # Closing EV3 device
> device.close()
> 

The expected result when executing the equivalent **MATLAB** code is:
[17, 0, 0, 0, 2, **48, 48, 49, 54, 53, 51, 52, 55, 48, 101, 53, 56**, 0, 0, 0, 0, 0, 0, 0]
[9, 0, 0, 0, 2, **69, 86, 51**, 0, 126, 126, 126, 126, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

Which converting the relevant UINT8s into ASCii characters is:
001653470e58
EV3

When I run the Python code above, I get:
[9, 0, 0, 0, 2, 69, 86, 51, 0, 126, 126, 126, 126, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[9, 0, 0, 0, 2, 69, 86, 51, 0, 126, 126, 126, 126, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

Which is the reply to the last direct command that was sent in the **MATLAB** session.
I can send the same command over and over and the reply doesn't change. It will only change if I open the device in **MATLAB**, send/receive a new message over there, and close it. Once I open the device using the Python implementation, no matter what messages I send to the device, I keep getting the last reply that I got in the **MATLAB** session.

Also, the **_open()_** and **_close()_** methods seem to be working. If I open the device in the Python environment, I cannot open it in the **MATLAB** one unless I close it in Python first. Vice-versa is also true.

As a side note, I tried the same type of Python code using **_pyusb_** (_libusb0_) and things worked as expected in that case. The problem with that package is that there isn't a **_close()_** method (or something equivalent). So, once the device is claimed, I can only open it with another application if I physically unplug and then plug back the USB cable. Not to mention the gruesome process of swapping back and forth between USB drivers...

Since the whole USB backend thing is somewhat an obscure science to me, I'm not sure if I'm doing something wrong in my code or if I missed a step (or two) in the installation process. All I did was run **_pip install hid_** and it worked (or at least I thought so...)

I've been Googling the hell out of this issue but I can't find any useful answers. Any help would be greatly appreciated!

Thanks,
Eduardo

**Comment by EduardoNigro** [comment]
Oh, by the way, I'm using Windows 10 and Python 3.9.

**Comment by prusnak** [comment]
What are the results when you change `reply = device.read(1024)[0:24]` to `reply = device.read(24)`? or to `reply = device.read(64)[0:24]`

**Comment by EduardoNigro** [comment]
Hello Pavol,

I still get the same repeat of whatever was the reply to the last command that was sent via the other application in **MATLAB**.

For instance, when the last command that I sent in **MATLAB** was a different one requesting the firmware version, that's what I get as a reply. Even when I made the changes you suggested. (See below)

> import hid
> 
> #### # OpeningEV3 device
> ID_VENDOR_LEGO = 0x0694
> ID_PRODUCT_EV3 = 0x0005
> device = hid.device()
> device.open(ID_VENDOR_LEGO, ID_PRODUCT_EV3)
> 
> #### # Getting EV3 brick serial number
> device.write(b'\n\x00\x00\x00\x00\x10\x00\xd3\x0c\x02\x10`')
> reply = device.read(64)[0:24]
> print(reply)
> 
> #### # Getting EV3 brick name
> device.write(b'\t\x00\x00\x00\x00\x08\x00\xd3\r\x08`')
> reply = device.read(24)
> print(reply)
> 
> #### # Closing EV3 device
> device.close()

The reply I get from the Python code above is:
[9, 0, 0, 0, 2, **86, 49, 46, 48, 57, 69**, 0, 126, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[9, 0, 0, 0, 2, **86, 49, 46, 48, 57, 69**, 0, 126, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

Which is the firmware version previously requested by the last **MATLAB** command. In ASCii:
V1.09E

I even changed the write command value to a list of integers or UINT8s with (or without) a padding of zeros to make it a 1024-byte package (which seems to be the size used in the **MATLAB** communication class), to no avail.

With my limited understanding of USB backends and their interfaces, it seems to me that both **_cython-hidapi_** and **MATLAB** are using **_hidusb.dll_**. At least, whenever I switch to the backend used by **_pyusb_** (using Zadig), that's what I see:

![image](https://user-images.githubusercontent.com/78741781/134844517-b3c447b9-3035-4e7e-b329-485622452b64.png)

Also, the interface that **MATLAB** uses seems to be related to the files **_hidapi-1.1.jar_** and **_mwhidapi.class_** located in the package folder and called in the USB communication class constructor:

>         function obj = CommUSB
>             % constructor
> 
>             % Adding EV3 to java path
>             jPath = javaclasspath('-dynamic');
>             ev3path = fileparts(which('legoev3'));
>             if ~ismember(ev3path, jPath)
>                 javaaddpath(ev3path);
>             end
>             hidjarpath = fullfile(ev3path, 'hidapi-1.1.jar');
>             if ~ismember(hidjarpath, jPath)
>                 javaaddpath(hidjarpath);
>             end
>             
>             try
>                 obj.USBHandle = mwhidapi;
>             catch 
>                 error('Connect USB Failed ');
>             end
>             obj.Status = 1;
>             
>         end

Is it possible that I have the incorrect backend for the **_cython-hidapi_** interface? How can I tell which one is being used?

Thanks again for your help.
Eduardo

**Comment by EduardoNigro** [comment]
Hey Pavol!

I finally figured it out! It was my mistake. Your code is fine and dandy.
It so happens that the byte sequence that one is sending as a direct command has to have a uint8 ZERO prepended to it. The **_send()_** method in **MATLAB** has it:

>         function send(obj, cmd)
>             % write
>             data = uint8(zeros(1, 1024));
>             data(2:1+length(cmd)) = cmd;
>             try
>                 obj.USBHandle.write(data);
>             catch
>                 error('Connect USB Failed ');
>             end
>         end

The line _data(2:1+length(cmd)) = cmd_ assigns the command starting at position 2 of the already pre-allocated with ZEROS data array. I overlooked that as I was working with the commands themselves and sending them to the EV3 brick directly.

As a side note, the first position of an array in MATLAB is 1 instead of 0.

Thank you for your help.
_"Just stare for long enough at your code and you will find the error..."_

Regards,
Eduardo
